ocrd_segment/repair.py

```diff
@@ -191,3 +144,158 @@ def process(self):
+                    LOG.warning('Region "%s" extends beyond Border of page "%s"',
+                                region.id, page_id)
+                    valid = False
+        for region in regions:
+            lines = region.get_TextLine()
+            for line in lines:
+                if not _child_within_parent(line, region):
+                    LOG.warning('Line "%s" extends beyond region "%s" on page "%s"',
+                                line.id, region.id, page_id)
+                    valid = False
+                if line.get_Baseline():
+                    baseline = LineString(polygon_from_points(line.get_Baseline().points))
+                    linepoly = Polygon(polygon_from_points(line.get_Coords().points))
+                    if not baseline.within(linepoly):
+                        LOG.warning('Baseline extends beyond line "%s" in region "%s" on page "%s"',
+                                    line.id, region.id, page_id)
+                        valid = False
+                words = line.get_Word()
+                for word in words:
+                    if not _child_within_parent(word, line):
+                        LOG.warning('Word "%s" extends beyond line "%s" in region "%s" on page "%s"',
+                                    word.id, line.id, region.id, page_id)
+                        valid = False
+                    glyphs = word.get_Glyph()
+                    for glyph in glyphs:
+                        if not _child_within_parent(glyph, word):
+                            LOG.warning('Glyph "%s" extends beyond word "%s" in line "%s" of region "%s" on page "%s"',
+                                        glyph.id, word.id, line.id, region.id, page_id)
+                            valid = False
+        return valid
+
+def _child_within_parent(child, parent):
+    child_poly = Polygon(polygon_from_points(child.get_Coords().points))
+    parent_poly = Polygon(polygon_from_points(parent.get_Coords().points))
+    return child_poly.within(parent_poly)
+
+def _plausibilize_group(regions, rogroup, mark_for_deletion):
+    wait_for_deletion = list()
+    reading_order = dict()
+    ordered = False
+    if isinstance(rogroup, (OrderedGroupType, OrderedGroupIndexedType)):
+        regionrefs = (rogroup.get_RegionRefIndexed() +
+                      rogroup.get_OrderedGroupIndexed() +
+                      rogroup.get_UnorderedGroupIndexed())
+        ordered = True
+    if isinstance(rogroup, (UnorderedGroupType, UnorderedGroupIndexedType)):
+        regionrefs = (rogroup.get_RegionRef() +
+                      rogroup.get_OrderedGroup() +
+                      rogroup.get_UnorderedGroup())
+    for elem in regionrefs:
+        reading_order[elem.get_regionRef()] = elem
+        if not isinstance(elem, (RegionRefType, RegionRefIndexedType)):
+            _plausibilize_group(regions, elem, mark_for_deletion)
+    for region in regions:
+        if (region in mark_for_deletion and
```

Comment:
This is in `O(1)` with `set` but in `O(n)` with list.

Reply:
Absolutely. It's a shame all our processors must be formulated so clumsily just because generateDS output is suboptimal. It will be hard to change that even after we start injecting custom code.